### PR TITLE
Use go method to calculate ssh key fingerprint

### PR DESCRIPTION
**Pull request by sapk**
Use ssh package to calculate the fingerprint in place of writing to a temporary file and relying on ssh-keygen to output the fingerprint.

This remove some i/o interaction and spinning up a sub-process. This is also needed for building a configurable rootless container because ssh-keygen try to find the running user (which is useless to calculate the fingerprint) 

https://godoc.org/golang.org/x/crypto/ssh#ParseAuthorizedKey
https://godoc.org/golang.org/x/crypto/ssh#FingerprintSHA256

### CODE REVIEW

**Comment by lunny** [comment]
It seems you can remove `ssh-keygen` totally if you change `SSHKeyGenParsePublicKey` on the same file.

**Comment by sapk** [comment]
@lunny It seems we implemented the two methods and only used it for built-in.  https://github.com/go-gitea/gitea/blob/744972ec8c9b1c445e5fc013762bda05337f5c3d/models/ssh_key.go#L287-L293

Should I do the same trick and implements both ? or should I clean SSHKeyGenParsePublicKey and replace it with SSHNativeParsePublicKey ?

File: models/ssh_key.go [hunk 1]
```diff
@@ -361,18 +361,11 @@ func checkKeyFingerprint(e Engine, fingerprint string) error {
```
Comment (techknowlogick) [inline models/ssh_key.go:362]: Please add Gitea copyright to header

Comment (sapk) [reply models/ssh_key.go:362]: done

**Comment by zeripath** [comment]
Hmm I think unless we're certain that go's support is the same as the native we could get into trouble assuming that go SSH and native fingerprints are the same. I don't know enough about go ssh to know if this is the case. The fact that we have this native switch for parsing makes me concerned that these may be different.

PS Why we can't get the fingerprint at the time of parsing?

**Comment by sapk** [comment]
@zeripath 
I think that go lib support of ssh type key should be the same as ssh-keygen now (the switch for parse date from 3 years ago) but maybe a conservative case would be to do the same and only use the go implementation for internal ssh. 
We can but since we have done it before by calling ssh-keygen we have done it in separate task.
If we want to do it in one action we should use the go implementation and don't use ssh-keygen. 
Looking at this part of the code, it would likely need to be refactor for other reason.

**Comment by sapk** [comment]
I separate the native go method to only to be use with built-in ssh. 
I added some test to validate both method (and refactor old test for SSHParsePublicKey to use sub-test logic). This way if a key is reported to fail with the native method we can easely reproduce by adding it to the key list to test.

**Review by lunny — APPROVED** [review]

**Review by techknowlogick — APPROVED** [review]

**Pull request merged**